plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/openfiles/OpenFileContext.java

```diff
@@ -0,0 +1,374 @@
+		this.parent = parent;
+		this.mainURI = uri;
+		this.temporary = isTemporary;
+		this.index = index;
+
+		this.mainResourceSet = createResourceSet();
+	}
+
+	protected XtextResourceSet createResourceSet() {
+		XtextResourceSet result = resourceSetProvider.get();
+		ResourceDescriptionsData.ResourceSetAdapter.installResourceDescriptionsData(result, index);
+		externalContentSupport.configureResourceSet(result, new OpenFileContentProvider());
+		return result;
+	}
+
+	/** Initialize the open file with the given content. */
+	protected void initOpenFile(int version, String content, CancelIndicator cancelIndicator) {
+		if (mainResource != null) {
+			throw new IllegalStateException("trying to initialize an already initialized resource: " + mainURI);
+		}
+
+		mainResource = (XtextResource) mainResourceSet.createResource(mainURI);
+		document = new XDocument(version, content);
+
+		try (InputStream in = new LazyStringInputStream(document.getContents(), mainResource.getEncoding())) {
+			mainResource.load(in, null);
+		} catch (IOException e) {
+			throw new RuntimeException("IOException while reading from string input stream", e);
+		}
+
+		resolveAndValidateOpenFile(cancelIndicator);
+	}
+
+	/** Initialize the open file based only on its URI, retrieving its content via EMF's {@link ResourceLocator}. */
+	protected void initOpenFile(CancelIndicator cancelIndicator) {
+		if (mainResource != null) {
+			throw new IllegalStateException("trying to initialize an already initialized resource: " + mainURI);
+		}
+
+		mainResource = (XtextResource) mainResourceSet.getResource(mainURI, true); // uses the EMF ResourceLocator
+		document = new XDocument(0, mainResource.getParseResult().getRootNode().getText());
+
+		resolveAndValidateOpenFile(cancelIndicator);
+	}
+
+	protected void refreshOpenFile(CancelIndicator cancelIndicator) {
+		// FIXME find better solution for updating unchanged open files!
+		TextDocumentContentChangeEvent dummyChange = new TextDocumentContentChangeEvent(document.getContents());
+		refreshOpenFile(document.getVersion(), Collections.singletonList(dummyChange), cancelIndicator);
+	}
+
+	protected void refreshOpenFile(@SuppressWarnings("unused") int version, // TODO add check using the version
+			Iterable<? extends TextDocumentContentChangeEvent> changes, CancelIndicator cancelIndicator) {
+
+		if (mainResource == null) {
+			throw new IllegalStateException("trying to refresh a resource that was not yet initialized: " + mainURI);
+		}
+		if (!mainResource.isLoaded()) {
+			throw new IllegalStateException("trying to refresh a resource that is not yet loaded: " + mainURI);
+		}
+
+		// TODO the following is only necessary for changed files (could be moved to #updateDirtyState())
+		ResourceSet resSet = getResourceSet();
+		for (Resource res : new ArrayList<>(resSet.getResources())) {
+			if (res != mainResource) {
+				res.unload(); // FIXME better way to do this? (unload is expensive due to re-proxyfication)
```

Comment:
With that approach: doesn't the mainResource then keep object references into the removed resources? Or is this unharmful, due to the code that follows (i.e. `mainResource.update(...)`)?

Reply:
Exactly: mainResource.update() should do a full reparse so all the objects are gone. There wouldn't be a need to to proxify any objects.
If it's a partial reparse, things become more annoying and we'd need to go for the full proxy dance. I'd rather prefer to avoid that. Do we disable partial parsing already in the AntlrParser via `isReparseSupported`?

Reply:
Yes, `N4JSSemicolonInjectingParser` overrides `isReparseSupported()` to always return `false`. But this is only N4JS. The code here has to work with all Xtext languages, if I'm not overlooking something.

Reply:
I'd rather focus on making this work with N4JS atm. The current change is already huge and therefore prone to take a few month before it has been brought into a shape that fits all languages (and other constrains). 
So in general: yes, you are right. On the other hand, unloading is so very expensive, that we probably want to avoid this at the expense of always doing a reparse(..)